libkmod/libkmod-index.c

```diff
@@ -135,6 +135,40 @@ enum node_offset {
 	INDEX_NODE_MASK = 0x0FFFFFFF, /* Offset value */
 };
 
+/* Write buffer to imitate FILE I/O optimization */
+struct wbuf {
+	char bytes[4096];
+	size_t len;
+	int fd;
+};
+
+static void wbuf_init(struct wbuf *buf, int fd)
+{
+	buf->len = 0;
+	buf->fd = fd;
+}
+
+static void wbuf_flush(struct wbuf *buf)
+{
+	if (buf->len > 0) {
+		write_str_safe(buf->fd, buf->bytes, buf->len);
+		buf->len = 0;
+	}
+}
+
+static void wbuf_write(struct wbuf *buf, const char *p, size_t len)
+{
+	if (buf->len + len > sizeof(buf->bytes)) {
+		wbuf_flush(buf);
+	}
+	if (len > sizeof(buf->bytes)) {
+		write_str_safe(buf->fd, p, len);
```

Comment:
Should probably check and adjust the buffer size so this doesn't trigger. Thus fewer code paths, greater coverage, etc.

Reply:
I'll adjust the code to always align with 4 KB writes, so the kernel wouldn't see a difference between a FILE implementation and ours. This check is required because index files could contain huge strings. In reality they won't, so this if-block is never entered. But no matter how large we make our buffer, we can always parse a string which is longer than our size on the stack. 

If you want to be funny with index files, nothing stops you from at least UINT_MAX chars in a string. The offsets would limit us, but there is no limit after a node starts ... :)

Reply:
> This check is required because index files could contain huge strings. In reality they won't, so this if-block is never entered

Sounds like we can move the check in the one(?) caller where it could trigger (for malicious input really). Might even make the function argument `const char buf[static 4096]` or alike, so compiler/static tools don't miss-report "omg this can overflow" scenarios.